File: Cpp/AdventOfCode/src/2016/d4_DecoyRoom.cpp
#include "2016/d4_DecoyRoom.h"

SOLUTION(2016, 4) {
    struct Room {
        std::vector<std::string> Parts;
        u32 Id = 0;
        std::string Checksum;
    };

    constexpr Room ParseRoom(std::string_view line) {
        auto s1 = Constexpr::Split(line, "[");
        auto s2 = Constexpr::Split(s1[0], "-");
        Room result;
        Constexpr::ParseNumber(s2.back(), result.Id);
        result.Checksum = s1[1].substr(0, s1[1].size() - 1);
        for (auto i = 0; i < s2.size() - 1; i++) {
            result.Parts.push_back(std::string(s2[i]));
        }
        return result;
    }

    constexpr std::string DecryptRoom(const Room& room) {
        auto rot = room.Id % 26;
        std::vector<std::string> decrypted;
        for (const auto& part : room.Parts) {
            std::string s;
            for (auto c : part) {
                s.push_back(static_cast<char>(((c - 'a' + rot) % 26) + 'a'));
            }
            decrypted.push_back(s);
        }

        return Constexpr::JoinVec(' ', decrypted);
    }

    bool IsReal(const Room& room) {
        std::string expectedChecksum = "";
        auto counts = std::array<s32, 26>{};

        for (const auto& part : room.Parts) {
            for (auto c : part) {
                counts[static_cast<size_t>(c - 'a')]++;
            }
        }

        std::vector<s32> sortedCounts;
        for (auto count : counts) {
            sortedCounts.push_back(count);
        }

        std::sort(sortedCounts.begin(), sortedCounts.end(), [](s32 lhs, s32 rhs) {return lhs > rhs; });

        for (auto i = 0; i < 5; i++) {
            auto expectedCount = sortedCounts[i];
            for (auto j = 0; j < counts.size(); j++) {
                if (counts[j] == expectedCount) {
                    expectedChecksum += static_cast<char>(j + 'a');
                    counts[j] = -1;
                    break;
                }
            }
        }

        return room.Checksum == expectedChecksum;
    }

    PART_ONE() {
        auto rooms = ParseLines(lines, ParseRoom);
        return Constexpr::ToString(std::accumulate(rooms.begin(), rooms.end(), 0, [](u32 prev, const Room& room) {
            return prev + room.Id * IsReal(room);
            }));
    }
    PART_TWO() {
        auto rooms = ParseLines(lines, ParseRoom);
        auto result = std::find_if(rooms.begin(), rooms.end(), [](const Room& room) {
            return IsReal(room) && DecryptRoom(room) == "northpole object storage";
            });
        return Constexpr::ToString(result->Id);
    }

    TESTS() {
        auto room = ParseRoom("qzmt-zixmtkozy-ivhz-343[abcde]");
        if (DecryptRoom(room) != "very encrypted name") return false;
        
        return true;
    }
}

File: Rust/rust_aoc/src/year_2016/d4_decoy_room.rs
pub fn part_one(lines: &Vec<String>) -> String {
    lines.iter().map(|line| Room::parse(line))
    .filter(|room| room.is_real())
    .map(|room| room.id)
    .sum::<u32>()
    .to_string()
}
pub fn part_two(lines: &Vec<String>) -> String {
    lines.iter().map(|line| Room::parse(line))
    .filter(|room| room.is_real())
    .map(|room| (room.decrypt(), room.id))
    .filter(|(name, _)| name == "northpole object storage")
    .map(|(_, id)| id)
    .sum::<u32>()
    .to_string()
}

struct Room {
    parts: Vec<String>,
    id: u32,
    checksum: String
}

impl Room {
    fn parse(line: &str) -> Self {
        let mut s1 = line.split("[");
        let parts = s1.next().unwrap().split("-").collect::<Vec<_>>();
        let checksum = s1.next().unwrap().strip_suffix("]").unwrap();
        let (last, elements) = parts.split_last().unwrap();
        let id = last.parse::<u32>().unwrap();
        let parts = elements.iter().map(|s| s.to_string()).collect();

        Self {parts, id, checksum: checksum.to_string() }
    }

    fn decrypt(&self) -> String {
        let rot = (self.id % 26) as u8;
        self.parts.iter().map(|p| p.as_bytes().iter().map(|c| rotate(c, rot)).collect::<String>()).collect::<Vec<_>>().join(" ")
    }

    fn is_real(&self) -> bool {
        let combined = self.parts.join("");
        let mut counts = ('a'..='z')
        .map(|c| (c, combined.chars().filter(|c2| c == *c2).count()))
        .collect::<Vec<_>>();
        //map of (a,4) (b,0)
        counts.sort_by(|(_, lhs), (_, rhs)| rhs.cmp(lhs));
        for index in 0..5 {
            if self.checksum.as_bytes()[index] as char != counts[index].0 {
                return false;
            }
        }

        true
    }
}

fn rotate(c: &u8, rot: u8) -> char {
    (((c - b'a' + rot) % 26) + b'a') as char
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn is_real_test() {
        assert!(Room::parse("aaaaa-bbb-z-y-x-123[abxyz]").is_real());
        assert!(Room::parse("a-b-c-d-e-f-g-h-987[abcde]").is_real());
        assert!(Room::parse("not-a-real-room-404[oarel]").is_real());
        
        assert!(!Room::parse("totally-real-room-200[decoy]").is_real());
    }
    #[test]
    fn decrypt_test() {
        assert_eq!("very encrypted name", Room::parse("qzmt-zixmtkozy-ivhz-343[abcde]").decrypt());
    }
}